pkg/nsx/services/subnetport/subnetport.go

```diff
@@ -522,3 +522,47 @@ func (service *SubnetPortService) IsEmptySubnet(id string, path string) bool {
 func (service *SubnetPortService) DeletePortCount(path string) {
 	service.SubnetPortStore.PortCountInfo.Delete(path)
 }
+
+func (service *SubnetPortService) GetMACByAttachmentID(vifs []mpmodel.VirtualNetworkInterface, attachmentID string) (string, error) {
+	filteredVIFs := []mpmodel.VirtualNetworkInterface{}
+	macAddresses := sets.New[string]()
+	for _, vif := range vifs {
+		if vif.LportAttachmentId != nil && *vif.LportAttachmentId == attachmentID {
+			filteredVIFs = append(filteredVIFs, vif)
+		}
+	}
+	// We observed in some cases, multiple VIFs are returned for the same attachment ID and their MAC addresses are the same.
+	// Not sure whether it's a NSX API bug or expected behavior. Whatever, we log a warning here, then continue because this may not break our logic, i.e. multiple vifs may have the same MAC address.
+	if len(filteredVIFs) > 1 {
+		log.Warn("Multiple VIFs found for attachment ID", "attachmentID", attachmentID, "filteredVIFs", filteredVIFs)
+	}
+	for _, vif := range filteredVIFs {
+		if vif.MacAddress != nil {
+			macAddresses.Insert(*vif.MacAddress)
+		}
+	}
+	if len(macAddresses) > 1 {
+		return "", fmt.Errorf("multiple MAC addresses found for attachment ID %s, filteredVIFs: %+v", attachmentID, filteredVIFs)
+	} else if len(macAddresses) == 0 {
+		return "", fmt.Errorf("no MAC address found for attachment ID %s, filteredVIFs: %+v", attachmentID, filteredVIFs)
+	}
+	return macAddresses.UnsortedList()[0], nil
+}
+
+func (service *SubnetPortService) GetAllVIFs() ([]mpmodel.VirtualNetworkInterface, error) {
+	pageSize := int64(1000)
+	var allVIFs []mpmodel.VirtualNetworkInterface
+	cursor := ""
+	for {
+		vifsPage, err := service.NSXClient.VifsClient.List(&cursor, nil, nil, nil, nil, &pageSize, nil, nil, nil)
```

Comment:
Can we use search API as i found https://{{policy}}/policy/api/v1/search/query?page_size=1000&query=resource_type:VirtualNetworkInterface seems to return the same result as VifsClient?
Then we can reuse the existing SearchResource function and save the results to store.
Even without search API, I feel using store index GetMACByAttachmentID can be more efficient.

Reply:
I see. By in the store, I'm not referring to the cache, I'm thinking about using a map or store instead of list for vifs; so that we can get the MAC by attachment id with O(1) complexity instead of O(n) by going through the vif list in GetMACByAttachmentID.

Reply:
Make sense, thanks for the suggestion, added a new VIF store.